common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -415,24 +490,65 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
     return mergeStatuses;
   }
 
   @Override
   public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
     if (logger.isDebugEnabled()) {
       logger.debug("register executor with RemoteBlockPushResolver {} local-dirs {} "
-        + "num sub-dirs {}", appId, Arrays.toString(executorInfo.localDirs),
-          executorInfo.subDirsPerLocalDir);
+        + "num sub-dirs {} shuffleManager {}", appId, Arrays.toString(executorInfo.localDirs),
+          executorInfo.subDirsPerLocalDir, executorInfo.shuffleManager);
+    }
+    String shuffleManagerMeta = executorInfo.shuffleManager;
+    if (shuffleManagerMeta.contains(":")) {
+      String mergeDirInfo = shuffleManagerMeta.substring(shuffleManagerMeta.indexOf(":") + 1);
+      try {
+        ObjectMapper mapper = new ObjectMapper();
+        MergeDirectoryMeta mergeDirectoryMeta = mapper.readValue(mergeDirInfo, MergeDirectoryMeta.class);
+        if (mergeDirectoryMeta.attemptId == ATTEMPT_ID_UNDEFINED) {
+          // When attemptId is -1, there is no attemptId stored in the ExecutorShuffleInfo.
+          // Only the first ExecutorRegister message can register the merge dirs
+          appAttemptsPathsInfo.computeIfAbsent(appId, id ->
+            new AppAttemptPathsInfo(appId, mergeDirectoryMeta.attemptId, executorInfo.localDirs,
+              mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir));
+        } else {
+          // If attemptId is not -1, there is attemptId stored in the ExecutorShuffleInfo.
+          // The first ExecutorRegister message from the same application attempt wil register
+          // the merge dirs in External Shuffle Service. Any later ExecutorRegister message
+          // from the same application attempt will not override the merge dirs. But it can
+          // be overridden by ExecutorRegister message from newer application attempt,
+          // and former attempts' shuffle partitions information will also be cleaned up.
+          AtomicBoolean newAttemptRegistered = new AtomicBoolean(false);
+          appAttemptsPathsInfo.compute(appId, (id, appAttemptPathsInfo) -> {
+            if (appAttemptPathsInfo == null || (appAttemptPathsInfo != null
+                && mergeDirectoryMeta.attemptId > appAttemptPathsInfo.attemptId)) {
+              appAttemptPathsInfo =
+                new AppAttemptPathsInfo(
+                  appId, mergeDirectoryMeta.attemptId, executorInfo.localDirs,
+                  mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir);
+              newAttemptRegistered.set(true);
+            }
+            return appAttemptPathsInfo;
+          });
+          if (newAttemptRegistered.get()) {
+            logger.warn("Remove shuffle partition info for {}_{} generated for former attempts",
+              appId, mergeDirectoryMeta.attemptId);
+            cleanupShufflePartitionInfo(appId, mergeDirectoryMeta.attemptId);
```

Comment:
`cleanupShufflePartitionInfo` iterates over the whole `partitions` map so we have increased the latency of registration. This cleanup can happen asynchronously. Please add a TODO and create a jira for it as well.